Fix signature rotation with rotate_with_page == false

DominikDostal opened the pull request:
## Description of the changes

Adds a counter rotation to a signature if it is added to a rotated pdf while rotate_with_page is false (See Issue https://github.com/MatthiasValvekens/pyHanko/issues/265)

## Caveats

Currently this solution adds the transformation matrix to the signature field first and then copies it to the appearence of the signature. This could be prevented if the already added /N field is reused instead of overwritten.
Note: /N or rather /AP was added twice even without my change, once in fields.py at `ap_dict[pdf_name('/N')] = pdf_out.add_object(ap_stream)` and once in cms_embedder.py where my change adds the /Matrix field

## Checklist

Please go over this checklist to increase the chances of your PR being worked on in a timely manner. Deviations are allowed with proper justification (see previous section).

 - [x] I have read the project's CoC and contribution guidelines.
 - [x] I understand and agree to the terms in the [Developer Certificate of Origin](https://developercertificate.org/) as applied to this contribution.
 - [x] My changes do not affect any public API or CLI semantics.
 - [ ] My PR adds regression tests (i.e. tests that fail if the bug fix is _not_ applied).
 -> I honestly dont know how to add this
 - [ ] All new code in this PR has full test coverage.
-> Also dont know how to ensure this

## Additional comments

I know that my solution is not perfect, since it creates a transformation matrix field that wont be used by the pdf in any way, but with the bug where 2 appearence objects are created, this was the only solution i could think of without also having to fix that.

I wouldnt mind if you reject this PR because of this, but i hope this can at least help you with fixing the bug.

MatthiasValvekens commented on pyhanko/sign/fields.py:
```diff
@@ -1665,6 +1665,23 @@ def append_signature_field(
+            while '/Parent' in pagetree_obj and ('/Type' not in pagetree_obj or pagetree_obj['/Type'] != '/Page'):
+                pagetree_obj = pagetree_obj['/Parent']
+            if '/Rotate' in pagetree_obj:
+                match pagetree_obj['/Rotate']:
```
Case matching is unfortunately not an option since pyHanko still supports Python 3.7+ ;)

DominikDostal replied:
Changed it to if/elif

MatthiasValvekens commented on pyhanko/sign/fields.py:
```diff
@@ -1665,6 +1665,23 @@ def append_signature_field(
+        matrix = None
+        if sig_field_spec.visible_sig_settings.rotate_with_page == False:
+            pagetree_obj = page_ref.get_object()
+            while '/Parent' in pagetree_obj and ('/Type' not in pagetree_obj or pagetree_obj['/Type'] != '/Page'):
```
That conditional looks doesn't look quite right. You want the closest ancestor with a `/Rotate` entry, no?

Also, perhaps keep track of seen pages here to prevent loops :).

DominikDostal replied:
Rotating the signature would only make sense if the page itself is rotated. Does pyHanko even add the signature to anything other than the page itself?

Im not sure if a `/Parent` field can ever become a loop, unless someone didnt understand what a parent is supposed to be in programming

MatthiasValvekens replied:
> Rotating the signature would only make sense if the page itself is rotated. Does pyHanko even add the signature to anything other than the page itself?

I thought you were trying to look for the closest ancestor with a `/Rotate` flag (that would be inherited down to the leaf node) to check if the page you're interested in was rotated? Or did I misinterpret your intent here?

> Im not sure if a /Parent field can ever become a loop, unless someone didnt understand what a parent is supposed to be in programming

Of course it can't be a loop in a well-formed PDF file, but getting stuck in an infinite loop on malformed ones is also not something we want ;).

DominikDostal replied:
> I thought you were trying to look for the closest ancestor with a `/Rotate` flag (that would be inherited down to the leaf node) to check if the page you're interested in was rotated? Or did I misinterpret your intent here?

No, you are totally right. I just assumed that there would be no rotation between signature and the actual page based on how pyHanko places the signature in the pdf. I will change it as soon as i find the time

> Of course it can't be a loop in a well-formed PDF file, but getting stuck in an infinite loop on malformed ones is also not something we want ;).

Who doesnt like infinite loops? You do have a point, you can never know how PDFs are generated. The reason i even opened the issue was because i had to deal with a PDF that was generated in portrait orientation and then rotated instead of directly generating it in landscape orientation

MatthiasValvekens commented on pyhanko/sign/fields.py:
```diff
@@ -1665,6 +1665,23 @@ def append_signature_field(
+                    case 270:
+                        matrix = (0.0, -1.0, 1.0, 0.0, 0.0, 0.0)
+            else:
+                matrix = (1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
```
Can we skip the matrix entry altogether if it isn't needed? That is, when the rotation is 0 or unset.

DominikDostal replied:
You are right, we can totally do that. I changed it

MatthiasValvekens commented on pyhanko/pdf_utils/content.py:
```diff
@@ -179,10 +184,12 @@ def __init__(
         resources: Optional[PdfResources] = None,
         box: Optional[BoxConstraints] = None,
         writer: Optional[BasePdfFileWriter] = None,
+        matrix: Optional[generic.ArrayObject] = None
```
Something's up with the typing, either here or elsewhere in the PR. Please double-check.

DominikDostal replied:
Changed them all to DictionaryObject

MatthiasValvekens commented on pyhanko/sign/signers/cms_embedder.py:
```diff
@@ -266,7 +266,10 @@ def apply(self, sig_annot, writer):
             )
-            sig_annot['/AP'] = stamp.as_appearances().as_pdf_object()
+            normalappearence = stamp.as_appearances()
+            if '/Matrix' in sig_annot:
```
This isn't legal IIRC. Will double-check, but I'm pretty sure the matrix entry only goes in the appearance stream dictionary.

Will also review in the context of your remark in the "Caveats" section.

MatthiasValvekens replied:
So, I just had another look at this part:

> Currently this solution adds the transformation matrix to the signature field first and then copies it to the appearence of the signature. This could be prevented if the already added /N field is reused instead of overwritten.
Note: /N or rather /AP was added twice even without my change, once in fields.py at `ap_dict[pdf_name('/N')] = pdf_out.add_object(ap_stream)` and once in cms_embedder.py where my change adds the /Matrix field.


I checked the usage sites you mentioned. They both exist to serve slightly different code paths: one is for creating new signature fields (that may or may not be signed in the same session), the other is to add signature appearances as part of a signing process. So they're both necessary and independently useful, but it's indeed quite possible that some workflows hit both of these.

Reusing the XObject stream pointed to by `/N` when generating the signature appearance looks good at first sight, but I'm not convinced it's the way to go. If you're signing a previously-prepared signature field in a workflow with multiple signers you want to avoid updating existing stream objects in-place for forensics reasons that I won't go into.

Now, what would be OK to do is to _copy_ the `/Matrix` entry (if any) from the "old" `/N` into the new one. That's OK because it doesn't require modifying a stream object. If my reading of your code is correct, that should also achieve what you want without putting nonstandard keys into the annotation dictionary :).

DominikDostal replied:
I honestly have no idea how that managed to be there and still work (or maybe i was hallucinating, who knows)
Is fixed now

MatthiasValvekens reviewed (changes requested):
Thanks! Needs a bit more work (esp. tests) and I still need to double-check what you wrote in the "Caveats" section, but this looks promising :).

DominikDostal commented:
Hey Matthias,

thanks for you feedback.

As i was trying to implement the suggestions and fix the mistakes i made, i had to admit to myself that i managed to be stupid and not know which parts of all my trial and error were actually needed for the solution.

The solution should have consisted of 3 parts:
1) Add a transformation matrix to the appearence stream to rotate the signature 
2) Change the box coordinates to be rotated
This was something i had done outside the pyHanko code, therefore it wasnt included in the PR.
I converted the box coordinates like this for `/Rotate 90`:
```python
rect = [generic.FloatObject(rect[0]), generic.FloatObject(rect[1]), generic.FloatObject((rect[3] - rect[1]) * -1 + rect[0]), generic.FloatObject(rect[2] - rect[0] + rect[1])]
```
3) Swap width and height of stamp
This was something that i happened to avoid with StaticStampStyle having the `inner_content_scaling=InnerScaling.STRETCH_FILL ` setting. As an example i printed out the parameter inside the `fit` method:
```
boxwidth:  96
boxheight:  532
inner_nat_width:  532
inner_nat_height:  96
x_pos:  0.0
y_pos:  0.0
x_scale:  0.18045112781954886
y_scale:  5.541666666666667
```
As i have already added 1) to the PR, i will try to implement the other 2 parts too.

**2):**
Thinking back on it, this seems to be a wrong calculation as i am just rotating the box with the upper left corner of the box staying on the same coordinate. I thinkt the best solution would be to not change the coordinates at all. Like if the box starts at 10|10 in portrait orientation, which is in the bottom left corner, then it should also be in the bottom left corner in landscape orientation

**3):**
I actually have no clue how im supposed to do this. Maybe you have an idea where this would fit?

MatthiasValvekens commented:
Hi Dominik,

Regarding transforming the box: this is surprisingly subtle. The **BBox** of a form XObject (in casu: the annotation's appearance stream) is in "form space" (i.e. the XObject's internal coordinate system), whereas the annotation's **Rect** is always in "default user space" (i.e. the page's coordinate system). The affine transformation matrix transforming from the former to the latter is what you put in the **Matrix** entry.

On the other hand, the **Rotate** entry transforms between the page coordinate system and what you see on the screen (let's call that "screen space"). So, if you want to specify your signature box in screen space (which seems to be the end goal here), you need to ensure that you apply the inverse transformation to **Rect** as well. 

... as I mentioned in the issue thread, it's not as straightforward as it first looks ;).

Does that answer your questions, or was it rather the maths that you needed some assistance with?
